boot/freeldr/bootsect/ntfs.S

```diff
@@ -0,0 +1,890 @@
+BootPartition:
+    .byte 1
+
+BootFlagSignature:
+    .word HEX(0aa55)    // BootSector signature
+
+// End of Bootsector
+// Next sector starts as sector 2, since boot sector is 
+// as a file under NTFS ($Boot, which has inode number 7)
+// and takes the two first clusters of the volume (which means
+// 16 sectors, checked on Windows and mkfs.ntfs from ntfsutils)
+
+#define ATTRIBUTE_DATA HEX(80)
+#define ATTRIBUTE_INDEX_ROOT HEX(90)
+#define ATTRIBUTE_INDEX_ALLOCATION HEX(A0)
+#define FILE_MAGIC HEX(454c4946)
+#define INDX_MAGIC HEX(58444e49)
+StartSearch:
+    // Compute MFT start sector
+    mov   eax, dword ptr [MftLocation]
+    mov   cl, byte ptr [SectsPerCluster]
+    movzx ecx, cx
+    mul   ecx
+    mov   dword ptr [MFTStartSector], eax
+
+    // Compute size of MFT entry in sectors
+    xor   ax, ax
+    mov   al, byte ptr [ClustersPerMftRecord]
+    test  al, al
+    js    NegativeOffset
+    mov   cx, word ptr [SectsPerCluster]
+    mul   cx
+    mov   word ptr [SectsPerMFT], cx
+    jmp   SearchNext
+    NegativeOffset:
+    // ClustersPerMftRecord is negative, so we need to perform 1 << (-ClustersPerMftRecord)
+    // to get the number of bytes needed for a MFT entry
+    not   al
+    inc   al
+    mov   cl, al
+    xor   ax, ax
+    inc   ax
+    shl   ax, cl
+
+    // But here want to store sectors, so we need to divide by BytesPerSector
+    xor   dx, dx
+    mov   cx, word ptr [BytesPerSector]
+    div   cx
+    mov   word ptr [SectsPerMFT], ax
+
+    SearchNext:
+    mov   bp, sp
+    sub   sp, HEX(10)
+
+    // Read Root Directory MFT
+    mov   ax, HEX(2000)
+    mov   es, ax
+    xor   bx, bx
+    xor   edx, edx
+    mov   eax, 5
+    call  ReadInode
+
+    xor   ax, ax
+    call  ExploreIndexRoot
+
+    push  eax
+    mov   ax, HEX(A00)
+    mov   es, ax
+    pop   eax
+    call  ReadInode
+
+DebugEnd:
+    xor   ax, ax
+    mov   ebx, HEX(30)
+    call  FindAttributeHdr
+    mov   si, ax
+    // Move to the attribute data
+    add   si, word ptr es:[si + HEX(14)]
+
+    // We don't support compressed, sparse or encrypted Freeldr
+    test  dword ptr es:[si + HEX(38)], HEX(4a00)
+    jnz   CompressedFreeldr
+
+    // Compute size in clusters
+    mov   eax, dword ptr es:[si + HEX(28)]
+    mov   cl, byte ptr [SectsPerCluster]
+    xor   edx, edx
+    div   ecx
+    mov   cx, word ptr [BytesPerSector]
+    movzx ecx, cx
+    xor   edx, edx
+    div   ecx
+    mov   edx, eax
+
+    push  ax
+    mov   si, offset msgLoading
+    call  PutChars
+    pop   ax
+
+    xor   ax, ax
+    mov   ebx, HEX(80)
+    call  FindAttributeHdr
+    mov   si, ax
+    add   ax, word ptr es:[si + HEX(20)]
+    xor   ecx, ecx
+    xor   bx, bx
+    push  FREELDR_BASE / 16
+    pop   fs
+    FreeLdrLoad:
+        pushad
+        call  ReadNonResidentAttribute
+        popad
+        mov   bx, fs
+        add   bx, HEX(100)
+        mov   fs, bx
+        xor   bx, bx
+        inc   ecx
+        cmp   ecx, edx
+        jbe   FreeLdrLoad
```

Comment:
Those three lines seem to cause a loading problem (takes ~ 40 seconds); @DarkFire01 and @archeYR have a patch for it that we're currently reviewing.

Reply:
@sdever Are you able to reproduce that problem I evoked above? Or does freeldr load fast for you?

Reply:
This has been resolved I think, I formatted a drive on XP and it worked as you’d expect. Unless I’m missing something can be resolved.